### PR TITLE
undo: PreUndo hook

### PR DESCRIPTION
**Pull request by NathanX-S**

### CODE REVIEW

**Comment by VeXan39231** [comment]
This would be nice

**Comment by robotboy655** [comment]
Do either of you have an actual use case for this or is this a "just because" proposition?

**Comment by NathanX-S** [comment]
@robotboy655 refunding props

**Comment by WilliamVenner** [comment]
Isn't the point of Garry's Mod to be as moddable as possible?

**Comment by GitSparTV** [comment]
> refunding props
Well what if undo didn't happen? In this case you need PostUndo

**Comment by NathanX-S** [comment]
What? PreUndo gets called right before the entities are removed but the undo is valid, this is what I need?

**Comment by WilliamVenner** [comment]
Maybe it should be moved into the `if ( count > 0 ) then` block, or better yet just make it shared and also call it in `Undo_FireUndo` on the server.

**Comment by WilliamVenner** [comment]
While we're here, can we bring some attention to: https://github.com/Facepunch/garrysmod/pull/1749 https://github.com/Facepunch/garrysmod/pull/1685

**Comment by NathanX-S** [comment]
@robotboy655

**Comment by Yogpod** [comment]
could be used to add a 'are you sure?' thing before undoing

**Comment by NathanX-S** [comment]
@Kefta

**Comment by GitSparTV** [comment]
> could be used to add a 'are you sure?' thing before undoing

This is a serverside hook tho.
Would you network the confirmation?
I used PlayerBindPress on gmod_undo to make Undo Lock

**Comment by NathanX-S** [comment]
@robotboy655

**Comment by Kefta** [comment]
Your use case wouldn't even work with this because of where it's placed. Nor would

> could be used to add a 'are you sure?' thing before undoing

because you don't have a return value. Change the hook to have a bool return value that can block the undo, add PostUndo before the return where it passed the undo object and count, then I'll merge it.

**Comment by Yogpod** [comment]
> Your use case wouldn't even work with this because of where it's placed. Nor would
> 
> 
> 
> > could be used to add a 'are you sure?' thing before undoing
> 
> 
> 
> because you don't have a return value. Change the hook to have a bool return value that can block the undo, add PostUndo before the return where it passed the undo object and count, then I'll merge it.

guess that's on me for not actually reading the code and just looking at the name

**Comment by NathanX-S** [comment]
paging @Kefta

File: garrysmod/lua/includes/modules/undo.lua [hunk 1]
```diff
@@ -356,7 +356,9 @@ function Do_Undo( undo )
 
 	if ( !undo ) then return false end
 	
-	hook.Run( "PreUndo", undo )
+	local call = hook.Run( "PreUndo", undo )
+
+	if ( !call ) then return end
```
Comment (Kefta) [inline garrysmod/lua/includes/modules/undo.lua:361]: This should be ``if ( call == false ) then return end``, othewise, all undos would be blocked. You should also add default GM funcs for PreUndo and PostUndo.

**Review by Kefta — CHANGES_REQUESTED** [review]

File: garrysmod/lua/includes/modules/undo.lua (hunk 1 above)
Comment (NathanX-S) [reply garrysmod/lua/includes/modules/undo.lua]: @Kefta What do you mean by default GM funcs?

Comment (Kefta) [reply garrysmod/lua/includes/modules/undo.lua]: > @Kefta What do you mean by default GM funcs?

Add GM:PreUndo (returning true) and GM:PostUndo to the base gamemode.

**Comment by NathanX-S** [comment]
paging @Kefta

**Comment by ThatLing** [comment]
more bloat hooks for no reason, just what we needed

**Comment by NathanX-S** [comment]
> more bloat hooks for no reason, just what we needed

a) irrelevant comment good shit
b) usecase was given, i personally only wanted one hook and was told to add another, cry somewhere else

**Comment by ThatLing** [comment]
you seem awfully upset about my single comment, maybe some timeout will do you some good

File: garrysmod/lua/includes/modules/undo.lua [hunk 2]
```diff
@@ -355,7 +355,11 @@ end
 function Do_Undo( undo )
 
 	if ( !undo ) then return false end
+	
+	local call = hook.Run( "PreUndo", undo )
```
Comment (Kefta) [inline garrysmod/lua/includes/modules/undo.lua:359]: Inline the if-statement instead of making this a local since it's only used in one place.

**Review by Kefta — CHANGES_REQUESTED** [review]

**Comment by NathanX-S** [comment]
@Kefta please accept

**Review by Kefta — APPROVED** [review]

**Comment by VeXan39231** [comment]
^ this is big

**Pull request merged**